controllers/openshift_controller.go

```diff
@@ -1022,7 +1024,16 @@ func (r *KataConfigOpenShiftReconciler) processKataConfigInstallRequest() (ctrl.
 	return ctrl.Result{}, nil
 }
 
-func (r *KataConfigOpenShiftReconciler) createExtensionMc(machinePool string) (ctrl.Result, error, bool) {
+// If the first return value is 'true' it means that the MC was just created
+// by this call, 'false' means that it's already existed.  As usual, the first
+// return value is only valid if the second one is nil.
+func (r *KataConfigOpenShiftReconciler) createExtensionMc(machinePool string) (bool, error) {
+
+	// In case we're returning an error we want to make it explicit that
+	// the first return value is "not care".  Unfortunately golang seems
+	// to lack syntax for creating an expression with default bool value
+	// hence this work-around.
+	var dummy bool
```

Comment:
There might be some functions where this could happen as well. What about making this a a function ?

```
// To be used in errors paths of functions that return (bool, error). This
// makes it explicit that the bool value is irrelevant in this case.
func dummyBool() bool {
	return false
}
```

Reply:
That's a good point but I'd strongly prefer doing this as a separate PR, if for no other reason then because I'm not aware of any other such place in the current code (though I agree that there probably are some).

BTW there are other similar refactors in my queue that I'd like to do (reconciliation return value, logging errors, getting rid of the "create resource manifest before we check if it's even going to be needed" pattern and possibly others) but I think it would be confusing to do those in a single PR, especially considering that they aren't prerequisites for what this PR tries to achieve.

Reply:
Thinking again about this trick. Since the bool return value is undefined in case of error, why not making that very explicit in the function and most of all in the callers by changing its signature to (*bool, error) ?

Reply:
Hm... that would make every caller branch both on non-nil error and nil bool, wouldn't it?  I'd consider that unfortunate, given that Go code is already 4-5 lines of error handling for each line that does something.  I think I like the `dummyBool()` idea better...

Reply:
> Hm... that would make every caller branch both on non-nil error and nil bool, wouldn't it?

No. The explicit contract would still be `As usual, the first return value is only valid if the second one is nil.`
i.e. if err is non-nil you go down the error path, else you can just dereference the bool pointer because it is guaranteed to be non-nil. No extra branch but an obvious crash if some caller doesn't honor the terms of the contract. This could be detected by a working CI.

Reply:
I believe such a simple protocol would be too strict to be generally useful.  Take the `updateNodeLabels()` function as an example whose return value is being dealt with in another commit in this PR.  Say it updates labelling of a bunch of nodes and then fails to do so for another node, maybe because updating its Node resource fails for any reason.  It will return an error but the first return value is still valid (it will be `true` because the function did update some nodes' labelling).  The caller will want to check the first value, error or not, since if any nodes' labelling was changed, it needs to follow up, plus it can reasonably assume that the Node resource update was transient so the node labelling job will likely be finished on the next reconciliation.

Also, using a contract that asks for dereferencing pointers without checking them first seems like asking for trouble with static code analysis.  I'm not sure about the current Golang but e.g. Rust would most likely complain, and if Golang doesn't at the moment it probably will in the future.  This is one of my personal gripes with static code analysis but people do seem to love linters.